File: leetcode_problems/python/lc1593_split_string_into_max_number_of_unique_substrings.py
class SolutionWrong:
    def maxUniqueSplit(self, s: str) -> int:
        uniq: set[str] = set()
        i = 0
        sub = ""
        while i < len(s):
            sub += s[i]
            if sub not in uniq:
                uniq.add(sub)
                sub = ""
            i += 1
        return len(uniq)


class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        def backtrack(i: int, uniq: set[str]) -> int:
            if i == len(s):
                return len(uniq)

            max_count = 0
            for j in range(i + 1, len(s) + 1):
                sub = s[i:j]
                if sub not in uniq:
                    uniq.add(sub)
                    max_count = max(max_count, backtrack(j, uniq))
                    uniq.remove(sub)
            return max_count

        return backtrack(0, set())


if __name__ == "__main__":
    s = "www z f v e d w fv h s ww"
    s = "wwwzfvedwfvhsww"
    # s = "ababccc"
    print(Solution().maxUniqueSplit(s))

File: leetcode_problems/cpp/lc1593_split_string_into_max_number_of_unique_substrings.cpp
#include <string>
#include <unordered_set>
#include <algorithm>
using std::string;
using std::unordered_set;

class Solution
{
public:
    int backtrack(int i, int n, string &s, unordered_set<string> &uniq)
    {
        if (i >= s.size())
        {
            return uniq.size();
        }

        int max = 0;
        for (int j = i + 1; j < n + 1; j++)
        {
            string sub = s.substr(i, j - i);
            if (!uniq.count(sub))
            {
                uniq.insert(sub);
                int imax = backtrack(j, n, s, uniq);
                max = std::max(max, imax);
                uniq.erase(sub);
            }
        }
        return max;
    }

    int maxUniqueSplit(string s)
    {
        int n = s.size();
        unordered_set<string> uniq;
        return backtrack(0, n, s, uniq);
    }
};
